vocode/streaming/output_device/twilio_output_device.py

```diff
@@ -3,66 +3,119 @@
 from vocode.streaming.utils.create_task import asyncio_create_task_with_done_error_log
+from vocode.streaming.utils.worker import InterruptibleEvent
 
 
-class TwilioOutputDevice(BaseOutputDevice):
+class ChunkFinishedMarkMessage(BaseModel):
+    chunk_id: str
+
+
+MarkMessage = Union[ChunkFinishedMarkMessage]  # space for more mark messages
+
+
+class TwilioOutputDevice(AbstractOutputDevice):
     def __init__(self, ws: Optional[WebSocket] = None, stream_sid: Optional[str] = None):
         super().__init__(sampling_rate=DEFAULT_SAMPLING_RATE, audio_encoding=DEFAULT_AUDIO_ENCODING)
         self.ws = ws
         self.stream_sid = stream_sid
         self.active = True
-        self.queue: asyncio.Queue[str] = asyncio.Queue()
-        self.process_task = asyncio_create_task_with_done_error_log(self.process())
 
-    async def process(self):
-        while self.active:
-            message = await self.queue.get()
+        self._twilio_events_queue: asyncio.Queue[str] = asyncio.Queue()
+        self._mark_message_queue: asyncio.Queue[MarkMessage] = asyncio.Queue()
+        self._unprocessed_audio_chunks_queue: asyncio.Queue[InterruptibleEvent[AudioChunk]] = (
+            asyncio.Queue()
+        )
+
+    def consume_nonblocking(self, item: InterruptibleEvent[AudioChunk]):
+        if not item.is_interrupted():
+            self._send_audio_chunk_and_mark(
+                chunk=item.payload.data, chunk_id=str(item.payload.chunk_id)
+            )
+            self._unprocessed_audio_chunks_queue.put_nowait(item)
+        else:
+            audio_chunk = item.payload
+            audio_chunk.on_interrupt()
+            audio_chunk.state = ChunkState.INTERRUPTED
+
+    def interrupt(self):
+        self._send_clear_message()
+
+    def enqueue_mark_message(self, mark_message: MarkMessage):
+        self._mark_message_queue.put_nowait(mark_message)
+
+    async def _send_twilio_messages(self):
+        while True:
+            try:
+                twilio_event = await self._twilio_events_queue.get()
+            except asyncio.CancelledError:
+                return
             if self.ws.application_state == WebSocketState.DISCONNECTED:
                 break
-            await self.ws.send_text(message)
+            await self.ws.send_text(twilio_event)
+
+    async def _process_mark_messages(self):
+        while True:
+            try:
+                # mark messages are tagged with the chunk ID that is attached to the audio chunk
+                # but they are guaranteed to come in the same order as the audio chunks, and we
+                # don't need to build resiliency there
+                await self._mark_message_queue.get()
+                item = await self._unprocessed_audio_chunks_queue.get()
+            except asyncio.CancelledError:
+                return
+
+            self.interruptible_event = item
+            audio_chunk = item.payload
+
+            if item.is_interrupted():
```

Comment:
the flow is (and was quite similar before):
- send each chunk as a `media` message, and alongside it a `mark` message
- this means that when the `media` message is played, we receive the `mark` message back indicating it was played
- on interrupt, send a `clear` message, as well as mark the interruptible event attached to each one of these chunks as interrupted (since we attach the same interruptible event to all of the chunks in `send_speech_to_output`) - the entrypoint to this is in `broadcast_interrupt`
- at this point, all of the `mark` messages get sent back to us, and we use them to track in `StreamingConversation` that the full utterance has been processed (either played or fully interrupted)

i'll get into this more in the synchronicity bit below, but there is a period between when Twilio recognizes the `clear` message and when we mark the interruption event when it's possible that chunks are marked as interrupted when they were actually played into the phone call. this is OK and is a small window, and it only affects the message that gets added to the transcript via `get_message_up_to`.

`mark` messages exist to track the processing of each chunk on the Twilio side